### ISSUE TITLE
Print preview is no longer WYSIWYG

### ISSUE THREAD

**Issue by mikfire** [Brewtarget/brewtarget#258]
The text browser was a pretty straight forward substitution for the web view, but it doesn't render quite the same way. The print preview no longer looks like what actually gets printed.

Qt 5.6 has hit my distro, so I could try the new web kit. I would be concerned that we are shipping against a very new Qt version.

Not sure if this is a bug, or not.

**Comment by rocketman768** [comment]
We need to fix all printing issues in 2.4.0.

**Comment by kapinga** [comment]
Is this still an issue? I just checked on the current 2.4.0 branch and the print preview looked a lot like the printed copy.

This ignores the fact that the 'Print Preview' is a resizable HTML window (so there's no pagination or margins), but otherwise it looked pretty good to me.

**Comment by pricelessbrewing** [comment]
Agreed with @kapinga This looks fine to me, although I think the html window that opens up should be set to a larger size. It was ~3" by 3" on my monitor and had no button to maximize to screen.

**Comment by xzfantom** [comment]
I want to bring this topic up, because it's in 2.4.0 project and milestone.
@mikfire is it still a problem?

**Comment by mikfire** [comment]
If I print to pdf, it doesn't look good. On the other hand, it looks pretty close to the print preview once the window is large enough. 

Both are really hard to parse. Maybe I should look at that and try to at least put in reasonable section headers and things.

I would consider this not to be a blocking issue.

**Comment by mattiasmaahl** [comment]
I'll give this a go. I have a few ideas on how to make it somewhat easy to understand and clean up the GUI at the same time.
My thinking here is to have 1 dialog to make all your printing and export from including previews. that includes the recipes, inventory and Brewday printouts.

**Comment by matty0ung** [comment]
If you want a starting point, have a look at https://github.com/Brewken/brewken/blob/develop/src/BtPrintPreview.h and https://github.com/Brewken/brewken/blob/develop/src/BtPrintPreview.cpp  Doesn't solve the WYSIWYG issue, but is a step towards unifying print preview.

**Comment by mattiasmaahl** [comment]
@matty0ung That looks good, although I could not get Brewken to build, either on my windows or my Debian, complained about UndoableAddOrRemoveList.h:79:104 was called with 0 arguments instead of the 8 required. 
One question though, is there a particular reason to why you're not using the QPrintPreviewDialog OR the QPrintPreviewWidget for displaying the printout?

**Comment by mattiasmaahl** [comment]
Getting the QPrintPreviewDialog running is not much code at all:
* Create the object
* connect the signals from the printer to a Slot you create for printing.
* Execute the preview and you're done basically.

But I'm in the midst in mocking up a suggestion on how I imagine the Printing dialog to look. give me a day or so to get that done and we can discuss the details then. 😄 
My thinking is that in the main menu you select Print. that will open a dialog with settings for printing where you can print a hardcopy, save as pdf or save as HTML. also options on what to include in the printout, i.e. recipe, brewday instructions inventory. One of them, all af them or selections of them.
let me draw a bit and update here then.

**Comment by mikfire** [comment]
It also isn't difficult -- I've the code somewhere -- to remember the size of the print preview window and simply set it to that when the window is created.

**Comment by mattiasmaahl** [comment]
Ok So I have been mocking about the printing dialog and have come up with this.
One unified Printing/save-as dialog that shows the exported content in relation to a page.
If it is WYSIWYG is yet to be determined as I have not yet had the chance to printout anything. (as Nothing in the dialog actually works yet 😊)

Feedback and suggestions is most welcome. I'm most certainly gonna continue this work next week.
Code can be found at: [https://github.com/mattiasmaahl/brewtarget/tree/feature/printpreview](https://github.com/mattiasmaahl/brewtarget/tree/feature/printpreview)
![bild](https://user-images.githubusercontent.com/1370239/119133545-b27cd500-ba44-11eb-96a1-02cae75ab29b.png)

**Comment by matty0ung** [comment]
Looks good.  Is it possible to add paper size select?  (ISTR default is Letter,  so it would be nice for folks outside North America and the Philippines to be able to switch to A4 without diving through too many menus.)

**Comment by matty0ung** [comment]
Also, it seems a bit confusing to say "Select printer" and then offer the Paper/PDF/HTML choice.  I know some applications offer "Print to PDF" as a sort of virtual printer.  But if we didn't want to do this, could we have a radio button for the Paper/PDF/HTML choice (less confusing to do one of these at a time I think).  Then the choice of printer could be greyed out unless Paper is selected.

Similarly, does the Portrait/Landscape choice make sense when the output is HTML?

**Comment by mattiasmaahl** [comment]
> Looks good. Is it possible to add paper size select? (ISTR default is Letter, so it would be nice for folks outside North America and the Philippines to be able to switch to A4 without diving through too many menus.)

Yes, paper size is a thing I was contemplating. Will add it into the form. May also add a configuration option ro select your own "default" settings for the printout.

**Comment by mattiasmaahl** [comment]
> Also, it seems a bit confusing to say "Select printer" and then offer the Paper/PDF/HTML choice. I know some applications offer "Print to PDF" as a sort of virtual printer. But if we didn't want to do this, could we have a radio button for the Paper/PDF/HTML choice (less confusing to do one of these at a time I think). Then the choice of printer could be greyed out unless Paper is selected.
> 
> Similarly, does the Portrait/Landscape choice make sense when the output is HTML?

Yes, checkboxes is a bit confusing, maybe a combobox for selecting output or as you say radiobuttons that grays out the printer-selector.

My initial thought was to enable multiole outputs at the same time. If you select all of them all would be saved. But that may not be a desired functionality.

I will make the changes next week and revert. I have a lot of work to do cleaning up an making tje recipeformatter play nice with the new printing form. Also finalize all connections and coding for the form it self.

One idea I had was to enable "multiselect" for the recipes to print more than one at the time? Any thoughts on that?

**Comment by matty0ung** [comment]
> One idea I had was to enable "multiselect" for the recipes to print more than one at the time? Any thoughts on that?
My general instinct is to start simple and then add more complexity/features if and when there's a need.  Personally I tend not to print out more than one recipe at a time, but I don't know about other users.

**Comment by mattiasmaahl** [comment]
Yes, that was my thought too, modifying the mainwindow::getRecipe(...) to return a list of selected recipes. Or maybe write a new one... don't know. But version 1 will have single print/save as for now, and future improvment for multiselect.

Was thinking of implementing a "template" system based on xml or yml too template the printouts giving the user oppotunity to change images an layout on them. But that for future development though, not gonna be done with the current work.

**Comment by mattiasmaahl** [comment]
Work is ongoing, have some ways to go yet but now it generates a printout that is WYSIWYG.
As I'm writing this I'm also preparing for my next part of this, adding templates so that the user can modidy what comes out on paper to some extent. (but more on that later)

for now you have this screenshot, more code is on it's way! and I also intend to write up some dev-doc as well :)


![bild](https://user-images.githubusercontent.com/1370239/121902167-bd6b1280-cd2f-11eb-9815-41dbb1934a89.png)

**Comment by matty0ung** [comment]
Looks good!

**Comment by mattiasmaahl** [comment]
ok, So I'm closing in on a PR again, just have a small issue with printPreview not working as expected when there is no printers installed on the system. Once I fix that I'll create a new PR.

**Comment by mattiasmaahl** [comment]
> 
> 
> It also isn't difficult -- I've the code somewhere -- to remember the size of the print preview window and simply set it to that when the window is created.
> 
> edit: But. If you are going to be in there. It might be nice to make printing the brewnotes an option? With a default of not printing them?

I believe I have implemented the printout to function this way, I also added the functionality to choose what inventory you want to print.

**Comment by mattiasmaahl** [comment]
> 
> 
> ok, So I'm closing in on a PR again, just have a small issue with printPreview not working as expected when there is no printers installed on the system. Once I fix that I'll create a new PR.

New PR up for grabs! have at it boys, merge or toss out! your choice!! 🤣

**Comment by matty0ung** [comment]
I will have a look soon (if @mikfire does not get there first).  I am just near merging the Recipe versioning stuff into Brewken, which is, amongst other things, a necessary prelude to backporting the new DB layer to Brewtarget.

**Comment by mattiasmaahl** [comment]
No worries!
Now this is built for develop, I briefly looked into porting this to stable/2.4.0 but did not have the time to fully asess the amount of work that would take.
